### ISSUE TITLE
[java] UseDiamondOperator: False positive with Guice TypeLiteral

### ISSUE THREAD

**Issue by davidboden** [pmd/pmd#6239]
**Affects PMD Version:**  7.18.0

**Rule:** [UseDiamondOperator](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#usediamondoperator)

**Description:**

Guice's TypeLiteral exists to capture type information. It forces you to create a subclass and thereby captures the generic type information for use at runtime. Removal of the type information by using the diamond operator changes the behaviour of the code, so PMD should not generate a warning and a suggestion to remove the type information and use the diamond operator.

**Code Sample demonstrating the issue:**

```java
import com.google.inject.TypeLiteral;

  protected Map<Key<?>, Object> computeBindings() {
    return Map.of(
        Key.get(new TypeLiteral<Optional<Long>>() {}, BusinessId.class),
        Optional.of(businessIdProvider.get()));
  }
```

**Expected outcome:**

PMD should allow the TypeLiteral to have the type information without generating a warning.

**Running PMD through:** Other: PMD plugin in IntelliJ

**Comment by mrclmh** [comment]
Other TypeLiterals that may be affected:
- javax.enterprise.util.TypeLiteral
- org.apache.commons.lang3.reflect.TypeLiteral

**Comment by oowekyala** [comment]
Should we just whitelist the name `TypeLiteral`? Or should we add a property of the rule with a list of types to ignore?

**Comment by mrclmh** [comment]
I overlooked "com.fasterxml.jackson.core.type.TypeReference".
Another possible hint for detection is the anonymous, empty subclass that you have to create in all variants I found.

**Comment by mrclmh** [comment]
Neither the example code above nor this example trigger the rule for me with 7.18 or 7.19:

```java
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.List;

public class JsonDeserializer {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    public List<String> deserializeToList(String json) throws JsonProcessingException {
        return MAPPER.readValue(json, new TypeReference<List<String>>() {});
    }
}
```

**Comment by mrclmh** [comment]
Correction:
I can now reproduce my example, but only with the PMD plugin in IntelliJ.
Not via CLI or in a unit test.

**Comment by UncleOwen** [comment]
@mrclmh Did you provide a complete auxclasspath when trying the CLI?

**Comment by mrclmh** [comment]
I can't check but I guess not.
